File: 838_push_dominoes.cpp
class Solution {
public:
    string pushDominoes(string a) {
        int i = 0, n = a.size();
        while (i < n) {
            if (a[i] != '.') { ++i; continue; }
            int j = i;
            while (j < n && a[j] == '.') ++j;
            char l = i ? a[i - 1] : '.', r = j < n ? a[j] : '.';
            if (l == 'R' && r == 'L') {
                int k = j - i;
                for (int x = 0; x < k / 2; ++x) a[i + x] = 'R', a[j - 1 - x] = 'L';
            } else if (l == 'R') fill(a.begin() + i, a.begin() + j, 'R');
            else if (r == 'L') fill(a.begin() + i, a.begin() + j, 'L');
            i = j;
        }
        return a;
    }
};

File: 838_push_dominoes.java
class Solution {
    public String pushDominoes(String s) {
        char[] a = s.toCharArray();
        int i = 0, n = a.length;
        while (i < n) {
            if (a[i] != '.') { i++; continue; }
            int j = i;
            while (j < n && a[j] == '.') j++;
            char left = i > 0 ? a[i - 1] : '.';
            char right = j < n ? a[j] : '.';
            if (left == 'R' && right == 'L') {
                int m = j - i;
                for (int k = 0; k < m / 2; k++) {
                    a[i + k] = 'R';
                    a[j - 1 - k] = 'L';
                }
            } else if (left == 'R') {
                Arrays.fill(a, i, j, 'R');
            } else if (right == 'L') {
                Arrays.fill(a, i, j, 'L');
            }
            i = j;
        }
        return new String(a);
    }
}
